### ISSUE TITLE
Add Docker testing and deployment

### ISSUE THREAD

**Issue by alex-l-kong** [angelolab/ark-analysis#502]
**Section 1: Design details**

**Relevant background**

Currently, each user has to build the Docker image locally before starting. This presents 2 problems:

- The setup on some users' computers may prevent the image from building correctly (ex. on certain M1 chip configurations)
- Because we have no mechanism in place to run our tests in a Docker environment, we have no way of catching bugs that specific users may be experiencing
- Building an image from scratch can take a very long time (ex. on the order of half an hour for pixel and cell clustering)

Addressing these issues requires we change our Travis testing and deployment in a two ways:

- Run our `pytest` commands in a Docker environment
- Deploy our image to Dockerhub with each new release so users can retrieve a pre-built image

**Design overview**

**NOTE: the following sections will be very much in flux**

_Testing in Docker_

Alternatively, Travis itself allows you to specify a `docker` service to run `docker` commands, however it doesn't provide a great way to run `pytest` in a containerized environment. Additionally, it's best to run this test outside of Travis to prevent the creation of too many environments.

The easiest way to do this is to use GitHub actions. We can use https://github.com/vanvalenlab/deepcell-tf/pull/575/files as a baseline.

_Deploying to Dockerhub_

We should include another stage in `.travis.yml` to do for this. The following link can guide us: https://github.com/vanvalenlab/deepcell-tf/blob/ad2b019fa912a7d417964429e4c4142abd0d3175/.travis.yml#L7. This template also prevents deployment on every single merge into `master`/`main` as it only deploys if the commit is tagged (similar to PyPI deployment).

We need to set up a Docker username and password if we don't already have one and store them in the Travis environment variables.

**Design list/flowchart**

After any commit:

- Send it to Travis, which runs the non-Docker and coverage tests
- Send it to GitHub actions which runs the Docker environment tests

After a tagged commit to `master`:

- Create a new Docker image and upload it to Dockerhub

**Required inputs**

N/A

**Output files**

N/A

**Section 2: Implementation details**

**Control flow**

In `.travis.yml`:

- Include another `stage` in the `jobs` section called `docker_deploy` which builds the Docker image on a tagged commit and sends it to Dockerhub

On Dockerhub:

- Create a specific Angelo lab account if we don't have one

On Travis:

Add the Dockerhub usernamd and password as environment variables for security

In `.github/workflows/actions`:

Add a step called `test_docker` which builds the Docker image (or uses a cached Docker image) and runs `pytest` inside it.

**Milestones and timeline**

It'll be easiest to get the Docker deployment set up first.

Then, I'll work on setting up a GitHub action to run the Docker environment tests.

Difficult to determine a specific timeline because of other projects and not being too familiar with setting either of these up. I imagine Docker deployment will be fairly easy but setting up the Docker tests might take longer.

**Comment by ngreenwald** [comment]
Looks good. I think first priority is the automatic deployment to docker hub. 

If we have a dependency issue with a new docker build, does dockerhub notify us somehow that the container failed to build? If so, I think the need for testing inside the container is reduced.
